### ISSUE TITLE
Djvu: No reader engine for this file or invalid file

### ISSUE THREAD

**Issue by ersi-dnd** [koreader/koreader#4469]
* KOReader version: v2019.01
* Device: Pocketbook Touch HD 3

#### Issue 

Error "No reader engine for this file or invalid file" upon opening djvu files. This applies to all djvu files on my device that the immediately previous version of Koreader had no problem with. I suspect that #1343 resurfaced.

**Comment by Frenzie** [comment]
Depends on what caused that, but I'd say that's unlikely. I imagine it's more likely to be a duplicate of #4467, which means it should be fixed in today's nightly. Could you give it a go?

**Comment by ersi-dnd** [comment]
You mean you already inserted some fix so that I can reupdate now to give it a go?

**Comment by Frenzie** [comment]
@NiLuJe did, but I haven't been able to try it myself yet. The "problem" is that it works on my Kobo either way. ;-)

**Comment by ersi-dnd** [comment]
First I reinstalled the previous version. Djvu files open fine. 

Then OTA update, which resulted in the same issue, "No reader engine for this file or invalid file"

**Comment by NiLuJe** [comment]
Might want to post the crash.log then, as, yeah, can't reproduce on Kobo (admittedly, I only checked a pdf).

**Comment by NiLuJe** [comment]
But I vaguely remember the djvu from the testsuite getting thumbnailed successfully after the fix on the emulator...

**Comment by NiLuJe** [comment]
Make sure you actually test the nightly (2019.01-2), and not the stable.

**Comment by ersi-dnd** [comment]
No problem with pdf files for me, just djvu.

Perhaps this is a relevant section from crash.log
`01/10/19-09:51:25 INFO  initializing for device PocketBook
01/10/19-09:51:25 INFO  framebuffer resolution: {
    ["h"] = 1448,
    ["w"] = 1072
}
ffi.load: libs/libfreetype.so.6
ffi.load: libs/libmupdf.so
ffi.load: libs/libwrap-mupdf.so
01/10/19-09:51:25 INFO  opening file /mnt/ext2/lugemisvara/historyofriseofh02bairuoft.djvu
01/10/19-09:51:26 WARN  cannot open document /mnt/ext2/lugemisvara/historyofriseofh02bairuoft.djvu error loading module 'libs/libkoreader-djvu' from file './libs/libkoreader-djvu.so':
	/mnt/ext1/applications/koreader/./libs/libk2pdfopt.so.2: undefined symbol: dewarpBuildPageModel_ex
01/10/19-09:51:30 INFO  Loading plugins from directory: plugins
01/10/19-09:51:30 INFO  Plugin  plugins/autosuspend.koplugin/main.lua  has been disabled.
ffi.load: sqlite3
ffi.load: z
ffi.load: sqlite3`

**Comment by NiLuJe** [comment]
Yep, that's #4467, meaning you're not running the latest nightly ;).

**Comment by ersi-dnd** [comment]
Is this one the latest nightly? http://build.koreader.rocks/download/nightly/v2019.01-2-gd02c340_2019-01-09/koreader-pocketbook-arm-obreey-linux-gnueabi-v2019.01-2-gd02c340_2019-01-09.zip

**Comment by NiLuJe** [comment]
Yes

**Comment by ersi-dnd** [comment]
Djvu files open now and, with luck, can be read. The app is so crashy now that I still have to revert to the earlier version.

Crashlog added
[crash.log](https://github.com/koreader/koreader/files/2744305/crash.log)

**Comment by Frenzie** [comment]
@NiLuJe Android log:

```
I/KOReader( 2461): error: [string "dl.lua"]:40: error opening /data/data/org.koreader.launcher/files/libs/libkoreader-djvu.so: dlopen failed: cannot locate symbol "pow" referenced by "libkoreader-djvu.so"...
I/KOReader( 2461): try to load module /data/data/org.koreader.launcher/files/libs/libkoreader-djvu.so
I/KOReader( 2461): error: [string "dl.lua"]:98: could not find library /data/data/org.koreader.launcher/files/libs/libkoreader-djvu.so
W/KOReader( 2461):  cannot open document /storage/sdcard/Download/djvu3spec.djvu error loading module 'libs/libkoreader-djvu' from file 'libs/libkoreader-djvu.so':
W/KOReader( 2461): 	dlopen failed: cannot locate symbol "pow" referenced by "libkoreader-djvu.so"..
```

**Comment by Frenzie** [comment]
Hold on, that looks suspiciously similar to #3493, except I was more or less randomly testing in Android 5.0 this time.

**Comment by Frenzie** [comment]
Hm, it looks like on my Android 5 emulator DjVu just doesn't work on any release. This might be a false alarm.

**Comment by Frenzie** [comment]
Yes, on various other Android versions it works. I'll have to check a bit further back to see when Android 5 broke, but it was at least over 4 months ago.

**Comment by Frenzie** [comment]
@NiLuJe Say, do you have an explanation for this? I discovered that DjVu won't load on v2019.01 if and only if you load a DjVu without loading a PDF first.

So start the program, load a DjVu, get this error:

```
01/10/19-18:48:12 WARN  cannot open document spec/front/unit/data/djvu3spec.djvu error loading module 'libs/libkoreader-djvu' from file 'libs/libkoreader-djvu.so':
	/home/frans/src/kobo/koreader/koreader-emulator-x86_64-linux-gnu/koreader/libs/libk2pdfopt.so.2: undefined symbol: dewarpBuildPageModel_ex
nil
```

Start the program, load a PDF, then a DjVu, everything's hunky dory.

This explains why I didn't catch it, but it's also wreaking some havoc on my plants to quickly write some extra anti-regression CI tests. (Hopefully temporary havoc, dinner first.) Because sure, I could just test DjVu first, but next time something like this happens it might be the other way around…

**Comment by NiLuJe** [comment]
@ersi-dnd: Define "earlier" ;).

Would that generally be the PB Touch HD instabilities reported in #2276?
Did it ever work?
Does it only happen with some specific workflows?

**Comment by NiLuJe** [comment]
@Frenzie : I'd say weird dynamic loader quirk related to lazy loading of weak symbols?

**Comment by NiLuJe** [comment]
@Frenzie: My first thought would be to go with introspection via objdump/readelf vs. any kind of runtime checks for this kind of stuff ;).

**Comment by NiLuJe** [comment]
@Frenzie : It's a bit unwieldy because you get a bunch of symbols that *will* actually get resolved at runtime because of the libc, but:

```
objdump -T koreader/libs/libk2pdfopt.so.2 | grep 'D  \*UND\*'

00000000  w   D  *UND*  00000000              __gmon_start__
00000000  w   D  *UND*  00000000              _ITM_deregisterTMCloneTable
00000000  w   D  *UND*  00000000              _ITM_registerTMCloneTable
00000000      D  *UND*  00000000              pthread_create
00000000      D  *UND*  00000000              dewarpBuildPageModel_ex
00000000      D  *UND*  00000000              pthread_join
```

```
readelf --dyn-syms koreader/libs/libk2pdfopt.so.2 | grep " UND " | grep " NOTYPE "

     0: 00000000     0 NOTYPE  LOCAL  DEFAULT  UND 
    22: 00000000     0 NOTYPE  WEAK   DEFAULT  UND __gmon_start__
    49: 00000000     0 NOTYPE  WEAK   DEFAULT  UND _ITM_deregisterTMCloneTab
    54: 00000000     0 NOTYPE  WEAK   DEFAULT  UND _ITM_registerTMCloneTable
    68: 00000000     0 NOTYPE  GLOBAL DEFAULT  UND pthread_create
   143: 00000000     0 NOTYPE  GLOBAL DEFAULT  UND dewarpBuildPageModel_ex
   162: 00000000     0 NOTYPE  GLOBAL DEFAULT  UND pthread_join
```

(That's on a broken nightly, of course ;)).

**Comment by Frenzie** [comment]
@NiLuJe Interesting suggestion. A quick search based based on the idea that someone probably already had that thought gave me this: https://gist.github.com/toojays/dffd9b806692a10649c1fcd146ee7866

And yes, there's probably nothing that can be done at runtime. Or well, not nothing per se, since you could also run a separate luajit instance for the relevant tests.

**Comment by ersi-dnd** [comment]
@NiLuJe  > Define "earlier" ;).

The immediately previous version. 

> Would that generally be the PB Touch HD instabilities reported in #2276?
> Did it ever work?
> Does it only happen with some specific workflows?

I guess I have been lucky to have been spared the instabilities. The first Koreader I installed on this device was one among the very last in 2018. I had no major stability problems.

Now, after upgrading, and after reverting, I have had major problems with djvu files. (But I did not revert a full clean way. I did not touch the koreader.app files. I will try reverting again by replacing those too.)

- Right bottom corner gesture does not work, regardless to which function it is configured to
- #4098 is back (it never occurred on this particular device before)
- Whether navigating in a file, in the file manager, or in the menus, the app keeps segfaulting often

So I am pretty sad for having upgraded and struggling now to get it to be usable again. Refraining from upgrading on my other devices for now.

**Comment by NiLuJe** [comment]
So, 2018.12?

Your device appears to b

**Comment by ersi-dnd** [comment]
@NiLuJe 
> As for #4098, does that happen on epubs?
Yes.

The first Koreader on this particular device that I ever installed and that worked well must have been 2018.11. I updated it to 2018.12 without any issues.

Right now, I am back on 2018.11 with a clean install and busy reinstating the settings manually one by one. It has intermittent stability issues. Some launches work fine for long, so that I can read and navigate menus. Other launches crash soon.

I discovered that upon first launches there was a consistent crash upon seventh (7th) touch in the menus. This crash-proneness eased when I, painstakingly, ticked off most of the plugins one by one. It also seems to help to open up the history, then exit history, and then it's possible to navigate the menus longer again, but this was inconsistent.

Anyway, almost happy again on 2018.11. Even #4098 is gone. Perhaps it's tied to one of my settings...

**Comment by NiLuJe** [comment]
@ersi-dnd: Would that 7th touch crash coincide with what should have been a flashing update, perhaps?

Keep us posted If you figure out which setting re-introduces #4098 on your end ;).

**Comment by ersi-dnd** [comment]
@NiLuJe 
> Would that 7th touch crash coincide with what should have been a flashing screen update, perhaps?

Quite possible. One of the things I do is tick off Flash buttons and Flash keyboard under Screen.

**Comment by Frenzie** [comment]
It might be easier to verify with page turns. I can't recall the default full refresh value otoh but it might be 6.

**Comment by NiLuJe** [comment]
Namely, I'm thinking newer PB are based on i.MX6 boards.

The new SDK is here: https://github.com/pocketbook/SDK_6.3.0
But what I'm thinking right now is that the kernel is much closer to what we're used to on Kobo & Kindle as far as the EPDC is concerned: https://github.com/pocketbook/Platform_MX6/blob/master/Kernel_631/include/linux/mxcfb.h

If their kernels aren't built with support for legacy i.MX5 ioctls, we're using wrong ioctls (especiall wait_for_complete). It wouldn't be unheard of: while Kobo has (so far) kept the legacy interface, Amazon burned all briudges (hell, even between the Oasis 2 and the PW4 they dropped debugging stuff).

**Comment by NiLuJe** [comment]
The tricky bit is I have no idea how to tell whether a PB device is based on those boards or not.

(Well... I have one terrible idea, the fb id, but it's not how we usually do this ;p).

**Comment by ersi-dnd** [comment]
Yes, the default refresh value is 6, but does startup count as number 1? And then 2 to open menus and this leaves only 5 touches to navigate the menus. Whereas I started counting when I tapped to open menus (=1) and the crash occurred consistently upon 7th touch.

Another thing, possibly related, is that when Koreader's screensaver setting is "leave as is" and I press the screenlock key on Pocketbook (or it goes to sleep), then I will have Pocketbook's screensaver image on the screen, not Koreader's. 

But when Koreader's screensaver setting is "use message" and the device goes to sleep/screenlock, then the screen ends up displaying Koreader's screensaver image, probably because Koreader has one more thing to do and this overcomes whatever Pocketbook does when it goes to sleep. I.e. there is as if a delay or a difficulty with Koreader's refresh on Pocketbook screen.

**Comment by NiLuJe** [comment]
Whether stuff increases the counter or not is a bit trickier than that, plus, some stuff will flash regardless of the counter ;).

What @Frenzie proposed is the best way to try it out (... at least for full-screen flashes. There's a remote possibility full flashes would work, but not partial ones >_<").

I can't say much about the SS behavior, as that entirely depends on how we interact with the underlying system, and I have no experience with PB. I can simply tell you that, yes, it's a huge mess on Kindle, so it's not out of the realm of possibilities that there's some race-y behavior involved ;).

**Comment by ersi-dnd** [comment]
@Frenzie 
> It might be easier to verify with page turns

You mean to try if it crashes when the number of page turns reaches the refresh point? Now that my settings are in place, I can safely do as many page turns as I want whichever way, keys or swipes or taps. There are only occasional hickups when the file is heavy, but it mostly gets past those too.

**Comment by NiLuJe** [comment]
And does the screen actually flash, or is nerfing that part of your usual settings, too? ;).

**Comment by ersi-dnd** [comment]
Yes, if you mean the screen going black for a moment. I just pushed some 30 pages ahead one by one, getting refresh flashes as supposed to.

**Comment by ersi-dnd** [comment]
Holding breath for some fifteen minutes, I updated again. No issues anymore scrolling the files and navigating the menus, stable enough to be usable. My settings were probably all retained. A few settings that I noticed being off were probably such that I had missed when reinstating my settings after the first troublesome update. 

The only issue I encountered was with the new dictionary downloader. On Pocketbook, the best I can say about it is that it's unresponsive. On Kobo (H2O version 1, with microSD slot) it is rather sluggish, the Download button does not respond visually, but the download still happens.
